### PR TITLE
Add sid gr model with validation on amzn beauty dataset

### PR DESCRIPTION
**Pull request by JacoCheung**
## Description
The related issue: #240 
In this PR, a new model is added: semantic id based decoder.
Some highlights:
1. We use mcore standard transformer block as the backbone of our decoder.However, our jagged input and beam search require special masks, as a result, we have to create padded dense attention mask and invoke the local impl of SPDA of mcore.
2. We implemented a simple beam search module, which is called during evaluation.
3. The model is not SOTA impl, and the convergence needs to be confirmed. Only amazon beauty dataset is tested. 
4. We now only support single GPU

## Checklist
- [x] Code Cleaning
- [x] README
- [x] CI pipeline ( integration test)

~~[CI:](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/41340134)~~
[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/41400712)

### CODE REVIEW

File: examples/sid_gr/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Semantic ID Generative Recommender Example
+
```
Comment (shijieliu) [inline examples/sid_gr/README.md:2]: we need to add some guidance in the readme

Comment (shijieliu) [reply examples/sid_gr/README.md:2]: we also need to update [project readme](https://github.com/NVIDIA/recsys-examples/blob/main/README.md#get-started)

**Comment by shijieliu** [comment]
@JacoCheung and how about we move all the ops and modules under commons?

File: examples/sid_gr/datasets/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from .disk_sequence_dataset import DiskSequenceDataset
```
Comment (shijieliu) [inline examples/sid_gr/datasets/__init__.py:1]: to make it consistent, how about make this folder as `dataset`?

Comment (JacoCheung) [reply examples/sid_gr/datasets/__init__.py:1]: `dataset` is somewhat confusing. Because we actually have also dataloader. The `data` seems more conventional. 

I would like to change the name under hstu to `data` . How do you think?

Comment (shijieliu) [reply examples/sid_gr/datasets/__init__.py:1]: data is not proper in my perspective. This is a [template](https://github.com/victoresque/pytorch-template) we can reference. It uses `data_loader` and `data` is used for storing training samples. And dlrmv3 is using [dataset](https://github.com/meta-recsys/generative-recommenders/tree/main/generative_recommenders/dlrm_v3)

Comment (JacoCheung) [reply examples/sid_gr/datasets/__init__.py:1]: `data` is used for storing training samples. I would say this is not quite true nowadays. This template seems out of date. 

Tried to explore several projects:

| proj                                                         | Folders            | Scope                                  |
| ------------------------------------------------------------ | ------------------ | -------------------------------------- |
| [Torchrec example 1](https://github.com/meta-pytorch/torchrec/tree/v1.3.0-rc1/examples/retrieval) | data               | DataLoader                             |
| [TorchRec example 2](https://github.com/meta-pytorch/torchrec/tree/v1.3.0-rc1/examples/bert4rec) | data  / dataloader | Data preprocessor / Dataset+DataLoader |
| [GRID](https://github.com/snap-research/GRID/tree/main/src/data) | data               | DataLoader+Dataset                     |
| [Megatron-core](https://github.com/NVIDIA/Megatron-LM/tree/main/megatron/core/datasets) | datasets           | Dataset+ DataLoader                    |
| [DLRMv3](https://github.com/meta-recsys/generative-recommenders/tree/main/generative_recommenders/dlrm_v3/datasets) | datasets           | Dataset                                |

My conclusion is either datasets or data is ok. For consistency, I'll rename sid_gr's `data` to `datasets`. (hstu's `dataset` to `datasets`). Is this OK to you?

Comment (shijieliu) [reply examples/sid_gr/datasets/__init__.py:1]: sure.

File: examples/sid_gr/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
```
Comment (shijieliu) [inline examples/sid_gr/README.md:2]: add a link to training readme

File: examples/sid_gr/modules/eval_metrics.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (shijieliu) [inline examples/sid_gr/modules/eval_metrics.py:1]: how about move modules inside commons?

File: examples/sid_gr/tests/test_beam_search.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+import pytest
```
Comment (shijieliu) [inline examples/sid_gr/tests/test_beam_search.py:1]: we can remove `sid_gr/ops/__init__.py` file

File: examples/sid_gr/modules/eval_metrics.py (hunk 4 above)
Comment (JacoCheung) [reply examples/sid_gr/modules/eval_metrics.py:1]: Let's do it after this release, along with the hstu refactoring.

Comment (shijieliu) [reply examples/sid_gr/modules/eval_metrics.py:1]: sure
